test/units/testsuite-64.sh

```diff
@@ -40,6 +40,57 @@ helper_check_device_symlinks() {(
     done < <(find "${paths[@]}" -type l)
 )}
 
+# Wait for a specific device link to appear
+# Arguments:
+#   $1 - device path
+#   $2 - number of retries (default: 10)
+helper_wait_for_dev() {
+    local dev="${1:?}"
+    local ntries="${2:-10}"
+    local i
+
+    for ((i = 0; i < ntries; i++)); do
+        test ! -e "$dev" || return 0
+        sleep .2
+    done
+
+    return 1
+}
+
+# Wait for the lvm2-pvscan@.service of a specific device to finish
+# Arguments:
+#   $1 - device path
+#   $2 - number of retries (default: 10)
+helper_wait_for_pvscan() {
+    local dev="${1:?}"
+    local ntries="${2:-10}"
+    local MAJOR MINOR pvscan_svc real_dev
+
+    # Sanity check we got a valid block device (or a symlink to it)
+    real_dev="$(readlink -f "$dev")"
+    if [[ ! -b "$real_dev" ]]; then
+        echo >&2 "ERROR: '$dev ($real_dev) is not a valid block device'"
+        return 1
+    fi
+
+    # Get major and minor numbers from the udev database
+    # (udevadm returns MAJOR= and MINOR= expressions, so let's pull them into
+    # the current environment via `source` for easier parsing)
+    source <(udevadm info -q property "$real_dev" | grep -E "(MAJOR|MINOR)=")
+    # Sanity check if we got correct major and minor numbers
+    test -e "/sys/dev/block/$MAJOR:$MINOR/"
+
+    # Wait n_tries*0.5 seconds until the respective lvm2-pvscan service becomes
+    # active (i.e. it got executed and finished)
+    pvscan_svc="lvm2-pvscan@$MAJOR:$MINOR.service"
+    for ((i = 0; i < ntries; i++)); do
+        ! systemctl -q is-active "$pvscan_svc" || return 0
```

Comment:
It looks still this only checks that the service is not active. Maybe, to make the service is actually invoked and finished, not only check active or not, but also get previous exist status or something?

Reply:
Or, `systemctl is-active` returns non-zero if it never triggered? I am not familiar that...

Reply:
Well, in this case the respective `lvm2-pvscan@x:y.service` remains in the `active` state after triggered (thanks to `RemainAfterExit=yes`), and gets into the `inactive` state when the LVM device is disconnected:

```
# lvm pvcreate ...
# lvm vgcreate ...
...
# systemctl status lvm2-pvscan@8:80.service
 ● lvm2-pvscan@8:80.service - LVM event activation on device 8:80
      Loaded: loaded (/usr/lib/systemd/system/lvm2-pvscan@.service; static)
      Active: active (exited) since Tue 2021-09-28 17:55:26 CEST; 838ms ago
        Docs: man:pvscan(8)
     Process: 1089 ExecStart=/usr/sbin/lvm pvscan --cache --activate ay 8:80 (code=exited, status=0/SUCCESS)
    Main PID: 1089 (code=exited, status=0/SUCCESS)
         CPU: 34ms
...
# iscsiadm ... --logout
# systemctl status lvm2-pvscan@8:80.service
○ lvm2-pvscan@8:80.service - LVM event activation on device 8:80
     Loaded: loaded (/usr/lib/systemd/system/lvm2-pvscan@.service; static)
     Active: inactive (dead)
       Docs: man:pvscan(8)
...
# iscsiadm ... --login
...
```

so I guess we should be in the clear here?
```

Reply:
Oh, `!` and `||`,,, complicated. Thanks.

Reply:
Yeah, it's a bit... weird, but otherwise you'd have to write `systemctl is-active ... && return 0 || :` to not trip over `set -e`. Not sure which one is less confusing...